README.md
Firmware SHA2017Badge
=====================

Read [project documentation](https://wiki.sha2017.org/w/Projects:Badge)
and get involved.

Project updates:

* [Wiki](https://wiki.sha2017.org/w/Projects:Badge)
* [Documentation](https://wiki.sha2017.org/w/Projects:Badge/Documentation)
* [Firmware](https://github.com/SHA2017-badge/Firmware)
* [PCB](https://github.com/SHA2017-badge/PCB)
* [Pad](https://pad.sha2017.org/p/badge)
* [Changelog](CHANGELOG.md)

Copyright (C) 2017 SHA2017 Badge team.
Based on template application for [Espressif IoT Development Framework (ESP-IDF)](https://github.com/espressif/esp-idf).
Copyright (C) 2016 Espressif Systems, licensed under the Apache License 2.0
as described in the file LICENSE.

[![Codacy Badge](https://api.codacy.com/project/badge/Grade/d73b02ee522942b8ae1244aa2ee62f43)](https://www.codacy.com/app/annejan/Firmware?utm_source=github.com&utm_medium=referral&utm_content=SHA2017-badge/Firmware&utm_campaign=badger)
[![Build Status](https://travis-ci.org/SHA2017-badge/Firmware.svg?branch=master)](https://travis-ci.org/SHA2017-badge/Firmware)

Debian prerequisites
--------------------

```
sudo apt-get install libncurses5-dev flex bison gperf python-serial libffi-dev libsdl2-dev libmbedtls-dev perl
```

Compiling and flashing
----------------------

```
git submodule update --init --recursive
source set_env.sh
make defconfig
make -j5
```

Optional configuration
-------------
```
make menuconfig
```

Compiling MicroPython
----------------------------------

```
make -j5 -C micropython/mpy-cross
make -j5 -C micropython/esp32/
```

Flashing MicroPython
----------------------------------

```
cd auto-flasher
cp ../micropython/esp32/build/application.bin firmware/sha2017-badge.bin
./auto_flash.pl /dev/ttyUSB0
```

Interacting via serial
----------------------
```
make monitor
```

Building on OSX
---------------

First you have to replace the linux toolchain:
* download the [mac toolchain](https://dl.espressif.com/dl/xtensa-esp32-elf-osx-1.22.0-61-gab8375a-5.2.0.tar.gz) from [this page](https://dl.espressif.com/doc/esp-idf/latest/get-started/macos-setup.html)
* rename the linux toolchain: `mv xtensa-esp32-elf xtensa-esp32-elf-linux`
* untar the mac toolchain: ` tar zxf ~/Downloads/xtensa-esp32-elf-osx-1.22.0-61-gab8375a-5.2.0.tar.gz`

Then install the `CP2102` [usbserial driver](https://www.silabs.com/products/development-tools/software/usb-to-uart-bridge-vcp-drivers)

In the `sdkconfig` file change this line:

    CONFIG_ESPTOOLPY_PORT="/dev/tty.SLAB_USBtoUART"

MicroPython
-----------
```
import badge
badge.eink_init()
badge.display_picture(0,-1)
import ugfx
ugfx.init()
ugfx.demo("HACKING")
ugfx.clear(badge.BLACK)
ugfx.thickline(1,1,100,100,badge.WHITE,10,5)
ugfx.box(30,30,50,50,badge.WHITE)
ugfx.string(150,25,"STILL","Roboto_BlackItalic24",badge.WHITE)
ugfx.string(130,50,"Hacking","PermanentMarker22",badge.WHITE)
len = ugfx.get_string_width("Hacking","PermanentMarker22")
ugfx.line(130, 72, 144 + len, 72, badge.WHITE)
ugfx.line(140 + len, 52, 140 + len, 70, badge.WHITE)
ugfx.string(140,75,"Anyway","Roboto_BlackItalic24",badge.WHITE)
ugfx.flush()
```
More info on the [MicroPython badge features](https://wiki.sha2017.org/w/Projects:Badge/MicroPython)

Badge Emulator
--------------
```
make -C micropython/unix
cd micropython/unix
./micropython ../../examples/Game\ of\ Life/game_of_life.py
```


CHANGELOG.md
# Change Log

## [Unreleased](https://github.com/SHA2017-badge/Firmware/tree/HEAD)

[Full Changelog](https://github.com/SHA2017-badge/Firmware/compare/sha2...HEAD)

**Closed issues:**

- vusb test in initial boot loader is too low [\#203](https://github.com/SHA2017-badge/Firmware/issues/203)
- sha2017\_ota.c: strncpy misuse \(potential buffer overflow\) [\#201](https://github.com/SHA2017-badge/Firmware/issues/201)
- sha2017\_ota.c: out-of-bounds read in sha2017\_read\_past\_http\_header [\#200](https://github.com/SHA2017-badge/Firmware/issues/200)
- SDCard Support [\#196](https://github.com/SHA2017-badge/Firmware/issues/196)
- Prevent sleeping when connected to power [\#193](https://github.com/SHA2017-badge/Firmware/issues/193)
- building on OSX [\#184](https://github.com/SHA2017-badge/Firmware/issues/184)
- Requested: test-app for vibration motor [\#182](https://github.com/SHA2017-badge/Firmware/issues/182)
- Leaving badge in launcher or installer accidentally drains battery fast [\#166](https://github.com/SHA2017-badge/Firmware/issues/166)
- Check if every badge\_\*.c component handles pre-init keypresses well. [\#146](https://github.com/SHA2017-badge/Firmware/issues/146)
- Freeing the HSPI bus breaks the VSPI config. [\#70](https://github.com/SHA2017-badge/Firmware/issues/70)

**Merged pull requests:**

- Update badge\_eink\_fb.c [\#209](https://github.com/SHA2017-badge/Firmware/pull/209) ([klondi](https://github.com/klondi))
- Add support for the decapitated final badge. [\#206](https://github.com/SHA2017-badge/Firmware/pull/206) ([basvs](https://github.com/basvs))
- Rewrite OTA update code [\#205](https://github.com/SHA2017-badge/Firmware/pull/205) ([basvs](https://github.com/basvs))
- Increase Vusb max from 5.5 to 5.7 volt. [\#204](https://github.com/SHA2017-badge/Firmware/pull/204) ([basvs](https://github.com/basvs))
- Notes on how to build the badge firmware on OSX [\#202](https://github.com/SHA2017-badge/Firmware/pull/202) ([nlitsme](https://github.com/nlitsme))
- add ugfx\_screen\_flipped bool to flip the screen. [\#195](https://github.com/SHA2017-badge/Firmware/pull/195) ([basvs](https://github.com/basvs))
- only update ota download percentage when screen is not busy [\#194](https://github.com/SHA2017-badge/Firmware/pull/194) ([basvs](https://github.com/basvs))
- Make OTA server and path configurable [\#191](https://github.com/SHA2017-badge/Firmware/pull/191) ([basvs](https://github.com/basvs))
- let ugfx.init\(\) use the nvs eink-type [\#189](https://github.com/SHA2017-badge/Firmware/pull/189) ([basvs](https://github.com/basvs))
- Update to newest micropython for installer fixes [\#188](https://github.com/SHA2017-badge/Firmware/pull/188) ([gavanfantom](https://github.com/gavanfantom))
- Update micropython to prevent 'too many open files' [\#187](https://github.com/SHA2017-badge/Firmware/pull/187) ([raboof](https://github.com/raboof))
- Update micropython-lib to get the urequests.py timeout feature [\#186](https://github.com/SHA2017-badge/Firmware/pull/186) ([gavanfantom](https://github.com/gavanfantom))
- bugfix in png parser \(bits per pixel\); added error-codes [\#185](https://github.com/SHA2017-badge/Firmware/pull/185) ([basvs](https://github.com/basvs))
- setup handlers before initializing. \(badge rev 0.1.0\) [\#181](https://github.com/SHA2017-badge/Firmware/pull/181) ([basvs](https://github.com/basvs))
- suppress compiler warning. [\#180](https://github.com/SHA2017-badge/Firmware/pull/180) ([basvs](https://github.com/basvs))
- remove unused variables. [\#178](https://github.com/SHA2017-badge/Firmware/pull/178) ([basvs](https://github.com/basvs))
- Enable assertions. [\#177](https://github.com/SHA2017-badge/Firmware/pull/177) ([basvs](https://github.com/basvs))
- add missing \_\_BEGIN\_DECLS / \_\_END\_DECLS [\#176](https://github.com/SHA2017-badge/Firmware/pull/176) ([basvs](https://github.com/basvs))
- add more doxygen updates [\#175](https://github.com/SHA2017-badge/Firmware/pull/175) ([basvs](https://github.com/basvs))
- misc minor updates. no changes in compiled code. [\#174](https://github.com/SHA2017-badge/Firmware/pull/174) ([basvs](https://github.com/basvs))
- Allow C++ compilers to include the badge headers. [\#172](https://github.com/SHA2017-badge/Firmware/pull/172) ([osresearch](https://github.com/osresearch))
- Add contributing guide [\#171](https://github.com/SHA2017-badge/Firmware/pull/171) ([raboof](https://github.com/raboof))
- Refactoring badge\_eink greyscale implementation [\#170](https://github.com/SHA2017-badge/Firmware/pull/170) ([basvs](https://github.com/basvs))
- workaround for the leds-disable. [\#169](https://github.com/SHA2017-badge/Firmware/pull/169) ([basvs](https://github.com/basvs))
- Reset the zsh hash after setting the PATH [\#168](https://github.com/SHA2017-badge/Firmware/pull/168) ([raboof](https://github.com/raboof))
- Clean and update sdkconfig.default and update IDF + micropython [\#165](https://github.com/SHA2017-badge/Firmware/pull/165) ([Roosted7](https://github.com/Roosted7))
- Support both eink types. [\#164](https://github.com/SHA2017-badge/Firmware/pull/164) ([basvs](https://github.com/basvs))
- BADGE\_EINK\_LUT\_FULL shouldn't do a partial update. [\#163](https://github.com/SHA2017-badge/Firmware/pull/163) ([basvs](https://github.com/basvs))
- Let micropython use the bpp powerdown manager [\#162](https://github.com/SHA2017-badge/Firmware/pull/162) ([Spritetm](https://github.com/Spritetm))
- split png header and data parser. [\#161](https://github.com/SHA2017-badge/Firmware/pull/161) ([basvs](https://github.com/basvs))
- Add raw screen buffer to badge\_eink\_update\(\) call. [\#160](https://github.com/SHA2017-badge/Firmware/pull/160) ([basvs](https://github.com/basvs))
- Makefile improvements [\#159](https://github.com/SHA2017-badge/Firmware/pull/159) ([basvs](https://github.com/basvs))

main/demo_ugfx.c
#include "sdkconfig.h"

#include <badge_button.h>
#include <badge_input.h>

#include <gfx.h>
#include <ginput/ginput_driver_toggle.h>

#include "demo_ugfx.h"

font_t roboto;
font_t robotoBlackItalic;
font_t permanentMarker;

const char* displayNames[] = {
    "Kartoffel",
    "Sebastius",
    "Boekenwuurm",
    "Attilla",
    "Stitch",
    "tsd",
    "Sprite_TM",
    "Underhand",
    "MarkusBec",
    "Roosted",
    "the_JinX",
    "realitygaps",
    "raboof"
};

#define numNames (sizeof(displayNames) / sizeof (const char*))

GEventToggle ptoggle;

void showDemo(uint8_t name, color_t front, color_t back) {
  ets_printf("Getting toggle status\n");
  if (!ginputGetToggleStatus(BADGE_BUTTON_UP, &ptoggle)) {
    ets_printf("Error getting toggle status\n");
  } else {
      if (ptoggle.on) {
        ets_printf("button was pressed\n");
      } else {
        ets_printf("button was not pressed\n");
      }
  }

  gdispClear(back);
  gdispDrawString(150, 25, "STILL", robotoBlackItalic, front);
  gdispDrawString(130, 50, displayNames[name], permanentMarker, front);
  // underline:
  gdispDrawLine(127 + 3, 50 + 22,
                127 + 3 + gdispGetStringWidth(displayNames[name], permanentMarker) + 14, 50 + 22,
                front);
  // cursor:
  gdispDrawLine(127 + 3 + gdispGetStringWidth(displayNames[name], permanentMarker) + 10, 50 + 2,
                127 + 3 + gdispGetStringWidth(displayNames[name], permanentMarker) + 10, 50 + 22 - 2,
                front);
  gdispDrawString(140, 75, "Anyway", robotoBlackItalic, front);
  gdispDrawCircle(60, 60, 50, front);
  gdispFlush();
}

void demoUgfx() {
  GListener pl;
  GEvent* event;

  ets_printf("Initializing gfx\n");
  gfxInit();
  ets_printf("Initialized gfx\n");

  GSourceHandle upHandle = ginputGetToggle(BADGE_BUTTON_UP);

  geventListenerInit(&pl);
  geventAttachSource(&pl, upHandle, GLISTEN_TOGGLE_ON|GLISTEN_TOGGLE_OFF);

  roboto = gdispOpenFont("Roboto_Regular12");
  robotoBlackItalic = gdispOpenFont("Roboto_BlackItalic24");
  permanentMarker = gdispOpenFont("PermanentMarker22");

  ets_printf("Clearing\n");

  gdispClear(Black);
  gdispFlush();

  while (1) {
    uint8_t i;
    for (i = 0; i < numNames; i++) {
      ets_printf("Make it White\n");
      showDemo(i, Black, White);
      ets_printf("Take 5\n");
      ets_printf("Waiting for 'up'\n");
      if ((event = geventEventWait(&pl, 5000))) {
        ets_printf("Got 'up' event\n");
      } else {
        ets_printf("Timeout without 'up' event\n");
      }

      ets_printf("Paint it Black\n");
      showDemo(i, White, Black);
      ets_printf("Take 5\n");
      ets_printf("Waiting for 'up'\n");
      if ((event = geventEventWait(&pl, 5000))) {
        ets_printf("Got 'up' event\n");
      } else {
        ets_printf("Timeout without 'up' event\n");
      }
    }
  }
}
